src/deepsparse/transformers/haystack/nodes.py

```diff
@@ -0,0 +1,367 @@
+    :param pooling_strategy: strategy for combining embeddings
+    :param emb_extraction_layer: f an int, the transformer layer number from
+        which the embeddings will be extracted. If a string, the name of last
+        ONNX node in model to draw embeddings from. If None, leave the model
+        unchanged. Default is -1 (last transformer layer before prediction head)
+    :param top_k: how many documents to return per query
+    :param progress_bar: if true displays progress bar during embedding.
+        Not supported by DeepSparse retriever nodes. Default is False
+    :param scale_score: whether to scale the similarity score to the unit interval
+        (range of [0,1]). If true (default) similarity scores (e.g. cosine or
+        dot_product) which naturally have a different value range will be scaled
+        to a range of [0,1], where 1 means extremely relevant. Otherwise raw
+        similarity scores (e.g. cosine or dot_product) will be used
+    :param embed_meta_fields: concatenate the provided meta fields and text
+        passage / table to a text pair that is then used to create the embedding.
+        This approach is also used in the TableTextRetriever paper and is likely
+        to improve  performance if your titles contain meaningful information for
+        retrieval (topic, entities etc.).
+    :param kwargs: extra arguments passed to EmbeddingExtractionPipeline
+    """
+
+    def __init__(
+        self,
+        document_store: BaseDocumentStore,
+        model_path: str,
+        batch_size: int = 1,
+        max_seq_len: int = 512,
+        pooling_strategy: str = "reduce_mean",
+        emb_extraction_layer: Union[int, str, None] = -1,
+        top_k: int = 10,
+        progress_bar: bool = False,
+        scale_score: bool = True,
+        embed_meta_fields: List[str] = [],
+        **kwargs,
+    ):
+        super(BaseRetriever).__init__()
+
+        self.document_store = document_store
+        self.model_path = model_path
+        self.batch_size = batch_size
+        self.max_seq_len = max_seq_len
+        self.pooling_strategy = pooling_strategy
+        self.emb_extraction_layer = emb_extraction_layer
+        self.top_k = top_k
+        self.progress_bar = progress_bar
+        self.scale_score = scale_score
+        self.embed_meta_fields = embed_meta_fields
+
+        _LOGGER.info(f"Init retriever using embeddings of model at {model_path}")
+        if self.progress_bar:
+            _LOGGER.warn(
+                "DeepSparseEmbeddingRetriever does not support progress bar, set "
+                "progress_bar to False"
+            )
+
+        self.embedding_encoder = DeepSparseEmbeddingEncoder(self, kwargs)
+
+    def train(*args, **kwargs):
+        raise NotImplementedError()
+
+    def save(*args, **kwargs):
+        raise NotImplementedError()
+
+    def load(*args, **kwargs):
+        raise NotImplementedError()
+
+
+class DeepSparseDensePassageRetriever(DensePassageRetriever):
```

Comment:
I wonder if this & the other retriever could be named more clearly to indicate what they do (at least to me the names don't mean anything haha).

Maybe `DualModelExtraction` and `SingleModelExtraction`? or something like that? Cause the differences are that the embedding retriever uses a single model and this one uses 2 models right? I think if we can capture those differences more in the names, it'll be easier for users to pick which one they should use.

Reply:
Afaict "Dense Passage Retrieval" and "Embedding Retrieval" are standard names for these approaches. Haystack (the document retrieval library I'm integrating with) implements a `DensePassageRetriever` class, and this is the `DeepSparse` version.

Although as an outside I totally agree. Something like `BiencoderRetriever` or the names you suggested would be much clearer

Reply:
Ooooh gotcha if they are standard names thats great!